### PR TITLE
Add NSG to CloudShell VNET

### PR DESCRIPTION
**Pull request by hima-gadupudi**
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

*
*
*

### CODE REVIEW

**Review by sdwheeler — APPROVED** [review]
Looks good to me.

**Comment by sdwheeler** [comment]
This PR updates the deployment templates documented in [Deploy Azure Cloud Shell in a virtual network with quickstart templates](https://learn.microsoft.com/azure/cloud-shell/quickstart-deploy-vnet). We need this PR merged so that we can update the documentation to match.

#sign-off

**Review by sdwheeler — CHANGES_REQUESTED** [review]
Commit the suggestion to remove the empty values. And update the prereq template per my email.

File: demos/cloud-shell-vnet/main.bicep [hunk 1]
```diff
@@ -67,10 +71,50 @@ resource containerSubnet 'Microsoft.Network/virtualNetworks/subnets@2021-08-01'
...
+          destinationPortRange: '*'
+          sourceAddressPrefix: containerSubnetAddressPrefix
+          destinationAddressPrefix: containerSubnetAddressPrefix
+          access: 'Deny'
+          priority: 100
+          direction: 'Inbound'
+          sourcePortRanges: []
+          destinationPortRanges: []
+          sourceAddressPrefixes: []
+          destinationAddressPrefixes: []
```
[the remark this answers is not in the document]
Comment (hima-gadupudi) [reply demos/cloud-shell-vnet/main.bicep]: Removed Empty arrays.

**Review by sdwheeler — APPROVED** [review]

**Comment by sdwheeler** [comment]
@anthony-c-martin Can this be merged?

File: demos/cloud-shell-vnet/main.bicep [hunk 2]
```diff
@@ -67,10 +71,46 @@ resource containerSubnet 'Microsoft.Network/virtualNetworks/subnets@2021-08-01'
         }
       }
     ]
+    networkSecurityGroup: {
+      id: nsgResourceId
```
Comment (anthony-c-martin) [inline demos/cloud-shell-vnet/main.bicep:75]:
```suggestion
      id: networkSecurityGroup.id
```

File: demos/cloud-shell-vnet/main.bicep [hunk 3]
```diff
@@ -67,10 +71,46 @@ resource containerSubnet 'Microsoft.Network/virtualNetworks/subnets@2021-08-01'
...
+  parent: networkSecurityGroup
+}
+
+resource networkSecurityGroup 'Microsoft.Network/networkSecurityGroups@2023-05-01' = {
+  name: nsgName
+  location: location
+  properties: {
+    securityRules: [
+      {
+        id: nsgResourceId
```
Comment (anthony-c-martin) [inline demos/cloud-shell-vnet/main.bicep:94]:
```suggestion
```

File: demos/cloud-shell-vnet/main.bicep [hunk 4]
```diff
@@ -67,10 +71,46 @@ resource containerSubnet 'Microsoft.Network/virtualNetworks/subnets@2021-08-01'
...
+          protocol: '*'
+          sourcePortRange: '*'
+          destinationPortRange: '*'
+          sourceAddressPrefix: containerSubnetAddressPrefix
+          destinationAddressPrefix: containerSubnetAddressPrefix
+          access: 'Deny'
+          priority: 100
+          direction: 'Inbound'
+        }
+        type: 'Microsoft.Network/networkSecurityGroups/securityRules'
```
Comment (anthony-c-martin) [inline demos/cloud-shell-vnet/main.bicep:107]:
```suggestion
```

File: demos/cloud-shell-vnet/main.bicep [hunk 5]
```diff
@@ -67,10 +71,46 @@ resource containerSubnet 'Microsoft.Network/virtualNetworks/subnets@2021-08-01'
...
+      id: nsgResourceId
+    }
+  }
+  dependsOn: [
+    networkSecurityGroup
+  ]
+}
+
+resource networkSecurityGroupDefaultRules 'Microsoft.Network/networkSecurityGroups/defaultSecurityRules@2023-05-01' existing = {
+  name: nsgName
```
Comment (anthony-c-martin) [inline demos/cloud-shell-vnet/main.bicep:84]: Is this name correct?

Comment (sdwheeler) [reply demos/cloud-shell-vnet/main.bicep:84]: @hima-gadupudi Can you respond to these suggestions and retest if you make changes?

File: demos/cloud-shell-vnet/main.bicep (hunk 2 above)
Comment (hima-gadupudi) [reply demos/cloud-shell-vnet/main.bicep:75]: This is variable created in variables area (line 47) and used it. If I replace as suggested, Bicep is throwing errors. I think keeping it like this is simple and reusable.

File: demos/cloud-shell-vnet/main.bicep (hunk 3 above)
Comment (hima-gadupudi) [reply demos/cloud-shell-vnet/main.bicep:94]: Same response as above.

File: demos/cloud-shell-vnet/main.bicep (hunk 5 above)
Comment (hima-gadupudi) [reply demos/cloud-shell-vnet/main.bicep:84]: name is correct.

**Pull request merged**